src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py

```diff
@@ -980,14 +980,18 @@ def _getPilotOptions(self, queue, **kwargs):
         else:
             self.log.info("DIRAC project will be installed by pilots")
 
-        # Pilot Logging defined?
-        pilotLogging = opsHelper.getValue("/Services/JobMonitoring/usePilotsLoggingFlag", False)
-        if pilotLogging:
-            pilotOptions.append("-z ")
+        # Pilot Logging defined? This enables the extended (possibly remote) logger.
+        # We need the URL and an optional flag which allows fine tuning on VO by VO basis:
+        pilotLogging = opsHelper.getValue("/Pilot/RemoteLogging", True)
+        remoteLoggerURL = opsHelper.getValue("/Pilot/RemoteLoggerURL", None)
+        # -z or --pilotLogging flags enable a remote Logger
+        if remoteLoggerURL is not None and pilotLogging:
+            pilotOptions.append("--pilotLogging")
+            pilotOptions.append("--loggerURL %s" % remoteLoggerURL)
```

Comment:
If we know (some of) the pilot configuration early and pass it from the `SiteDirector` we can log early, even before pilot is installed and definitely before pilot commands run. Currently `pilot.json` contains only part of the `Operations`, namely `Defaults` and `Setup(s)`. If we wish to be able to switch logging on and off on a VO by VO basis we would need to extend this part of `pilot.json`. This can be done. A question which remains is when we want to use this information. If we want to log early we would need to read and analyse `pilot.json` file in the wrapper (so practically in `SiteDirector` and not in pilot commands). 

Another question: How would submission to clouds (`CloudDirector`) fits into this scenario ? To be independent of Directors we would need to log from the pilot, it mean logging late.

Reply:
Hi,
There is one issue we missed when we discussed remote logging from pilots.
When we want to log based entirely on `pilot.json `(in the case where `SiteDirector`is not a job submitter) we need a way to get a `VO` within the pilot to be able to verify if the `VO` is willing to log remotely. We can get a proxy, but all the nice utilities are only available in Dirac: `getProxyInfo()` or even better `getVOfromProxyGroup()`. Do we want to port this to `PilotTools` ?

Reply:
There is no problem to export more data to `pilot.json` from the CS. The problem is to get the `VO` name early (before Dirac is installed) from a running pilot `(pilotTools)` without duplicating the functionality present in Dirac, i.e in `getVOfromProxyGroup()` and Dirac-specific modules imported there.

Reply:
and if the `vo` parameter is passed with a `dirac-pilot` switch (from SiteDirector), will it start logging earlier?

Reply:
Yes, as we are doing it now, a difference is that rather that passing the `url` and a logging level we would pass the `VO` and decide what to do within the pilot, based on a json file. The problem is what we do if there is no` SiteDirector`. How will we get the `VO` in a multi-VO environment? 

What does it look like  in a case of a single VO environment ? Are the pilot parameters in `Defaults` ? (if yes, then at least we can cope with logging in a single VO case and no `SiteDirector` since a `Defaults` in `pilot.json` is basically a subset of `Defaults` in the  `CS`.

In a multi-VO case we need a VO, and if it is not provided by (a non existent) `SiteDirector` then we need to get it from a proxy. Getting the `VO` from a proxy was my primary concern.

Reply:
To be precise: we need a `VO` to be able to file-tune the logging on a` VO` by `VO` basis.

Reply:
A few points:
- in the long-ish run, we will most probably remove the concept of "Setup" -- this is already getting out of the code (see PRs 6164 and 6419, plus some related discussions in https://github.com/DIRACGrid/DIRAC/issues/5287), so I would say that we should consider like every parameter is in "Defaults".
- there is at the moment an issue, it seems to me, that the PilotCStoJSONSynchronizer does not add to the pilot.json any info about the VO, see e.g. what's produced in DIRAC certification (https://lbcertifdirac70.cern.ch/pilot/pilot.json) which is a multi-VO installation. 
- In case we update the format of pilot.json, e.g.including the VO and removing the Setup, we need to be careful not to break the pilot running on older versions
  - partly related, we could simply move the whole CS content to a JSON file. The reason this was NOT done since the beginning was to avoid exposing a content that might contain sensitive information, but maybe some part could be removed in a sort of post-conversion... what do you think?
- an easy way to get the `VO` from a proxy would be using `voms-proxy-info`, e.g. 
  ```bash
  $ voms-proxy-info -file /tmp/x509up_u49429 -vo
  lhcb
  ```
  but this of course relies on `voms-proxy-info` being on the WN before DIRAC is installed, or using a cvmfs-deployed version, like the one in dirac.egi.eu/client/diracos/. This would be OK most probably for several LCG resources but not necessarily for all resources (and on HPCs we might be in trouble). Base `openssl` is not enough.
  We can of course rely on "sensible defaults" if we fail to get the VO from the proxy (and ~soon from the token).

Reply:
I read the proposed CS change document. From what I understand, the pilot settings will stay where they are, only other VO-specific setting will be moved to another root section (hopefully not called SenSettings), right ?

If this is the case, we can look into dumping pilot section of `Operations` into `pilot.json`. Not sure where, though.
My `pilot.json` on my v7r3 box does not contain any `VO` -  specific data either. Layout  in `tests/CI/pilot.json` looks quite complete, with default sections only. The section contains, among other things `GenericPilotGroup` which is VO specific, so a good candidate for a VO section.

There is a tempting minimal solution - use a 
`"Logging": {
        "LoggingType":"LOCAL_FILE", 
        "LocalOutputFile":"myFile"
      },`
section to hold only logging related info, I'm not a great fan of this. It look artificial, there is no logging section in the CS.

Should the new section come after the outer Defaults ?

```json 
    "Defaults": {
      "Commands": {
        "defaultList": [
          "CheckWorkerNode",
          "InstallDIRAC",
          "ConfigureBasics",
          "CheckCECapabilities",
          "CheckWNCapabilities",
          "ConfigureSite",
          "ConfigureArchitecture",
          "ConfigureCPURequirements"
        ],
        "Defaults": [
          "CheckWorkerNode",
          "InstallDIRAC",
          "ConfigureBasics",
          "CheckCECapabilities",
          "CheckWNCapabilities",
          "ConfigureSite",
          "ConfigureArchitecture",
          "ConfigureCPURequirements",
          "LaunchAgent"
        ]
      },
      "ConfigurationServer": "VAR_CS",
      "GenericPilotGroup": "dirac_pilot",
      "GenericPilotDN": "VAR_USERDN"
    }
    "VOSetting":{"gridpp":{
      "GenericPilotGroup": "gridpp_pilot",
      "GenericPilotDN": "/C=UK"
      "RemoteLogging" : "True"
    }
```

etc.

Reply:
Well, right now from a CS section like this (taken from the certification CS, removed the non-meaningful stuff):

```
DIRAC-Certification
{
  Pilot
  {
    CheckVersion = True
    Version = 8.1.0a3
    pilotFileServer = lbcertifdirac70.cern.ch
    Commands
    {
      Test = GetPilotVersion
      Test += CheckWorkerNode
      Test += InstallDIRAC
      Test += ConfigureBasics
      Test += ConfigureCPURequirements
      Test += ConfigureArchitecture
      Test += CheckCECapabilities
      Test += LaunchAgent
    }
    Extensions = None
    pilotRepoBranch = devel
  }
}
dteam
{
  DIRAC-Certification
  {
    Pilot
    {
      GenericPilotGroup = dteam_pilot
      GenericPilotDN = /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=fstagni/CN=693025/CN=Federico Stagni
    }
    Cloud
    {
      GenericCloudGroup = dteam_pilot
      GenericCloudDN = /DC=ch/DC=cern/OU=computers/CN=lbcertifdirac70.cern.ch
    }
  }
}
gridpp
{
  DIRAC-Certification
  {
    Pilot
    {
      GenericPilotGroup = gridpp_pilot
      GenericPilotDN = /C=UK/O=eScience/OU=Imperial/L=Physics/CN=daniela bauer
    }
    Cloud
    {
      GenericCloudGroup = gridpp_pilot
      GenericCloudDN = /DC=ch/DC=cern/OU=computers/CN=lbcertifdirac70.cern.ch
    }
  }
}
```

We produce, in an artificial way, the pilot.json:
```json
{
  "Setups": {
    "DIRAC-Certification": {
      "CheckVersion": "True",
      "Version": "8.1.0a3",
      "pilotFileServer": "lbcertifdirac70.cern.ch",
      "Extensions": "None",
      "pilotRepoBranch": "devel",
      "Commands": {
        "Test": [
          "GetPilotVersion",
          "CheckWorkerNode",
          "InstallDIRAC",
          "ConfigureBasics",
          "ConfigureCPURequirements",
          "ConfigureArchitecture",
          "CheckCECapabilities",
          "LaunchAgent"
        ]
      }
    }
  },
  "GenericPilotDNs": [],
  "DefaultSetup": "DIRAC-Certification",
}
```

Where as you can see we ignore the VO-level information: for example, we could, in theory, have a list of Pilot Commands defined at the VO level, but they would be ignored. I think this is just a bug! For completeness, in `PilotCStoJSONSynchronizer`, everything starts with `options = gConfig.getOptionsDict("/Operations/%s/Pilot" % setup)` so a section like `Operations/gridpp/DIRAC-Certification/Pilot` is ignored.

I think we should start by simply using the `diraccfg` function of converting to json, so e.g. the cfg above would become:

```
{
  "DIRAC-Certification": {
    "Pilot": {
      "CheckVersion": "True",
      "Version": "8.1.0a3",
      "pilotFileServer": "lbcertifdirac70.cern.ch",
      "Extensions": "None",
      "pilotRepoBranch": "devel",
      "Commands": {
        "Test": "GetPilotVersion, CheckWorkerNode, InstallDIRAC, ConfigureBasics, ConfigureCPURequirements, ConfigureArchitecture, CheckCECapabilities, LaunchAgent"
      }
    }
  },
  "dteam": {
    "DIRAC-Certification": {
      "Pilot": {
        "GenericPilotGroup": "dteam_pilot",
        "GenericPilotDN": "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=fstagni/CN=693025/CN=Federico Stagni"
      },
      "Cloud": {
        "GenericCloudGroup": "dteam_pilot",
        "GenericCloudDN": "/DC=ch/DC=cern/OU=computers/CN=lbcertifdirac70.cern.ch"
      }
    }
  },
  "gridpp": {
    "DIRAC-Certification": {
      "Pilot": {
        "GenericPilotGroup": "gridpp_pilot",
        "GenericPilotDN": "/C=UK/O=eScience/OU=Imperial/L=Physics/CN=daniela bauer"
      },
      "Cloud": {
        "GenericCloudGroup": "gridpp_pilot",
        "GenericCloudDN": "/DC=ch/DC=cern/OU=computers/CN=lbcertifdirac70.cern.ch"
      }
    }
  }
}
```

But this means quite some modifications to the Pilot (pilotTools.py) code to make sure that both formats are understood.

Reply:
I see. You have omitted `Setups:` at the top, is it intentional ? What about `Operations/Defaults/Pilot` ? This is a place when we can set remote logging to `False` for all `VO`s and re-enable it down in `VO`-specific sections. It holds for all `Setups,` I believe.

Shouldn't we arrange for a short face-to-face ;-) Zoom meeting to clarify things before coding ?

Reply:
> I see. You have omitted `Setups:` at the top, is it intentional ? What about `Operations/Defaults/Pilot` ? This is a place when we can set remote logging to `False` for all `VO`s and re-enable it down in `VO`-specific sections. It holds for all `Setups,` I believe.

I intentionally omitted it, and I am claiming it could be always omitted, as I don't see it useful.
`Operations/Defaults/Pilot` is valid, and it has nothing to do with Setups. I agree that it is a place where we can set remote logging to `False` for all `VO`s and re-enable it down in `VO`-specific sections. This is independent from the Setups.


> Shouldn't we arrange for a short face-to-face ;-) Zoom meeting to clarify things before coding ?

Definitely. I will write you mail.

Reply:
As I'm looking into the `diraccfg` command it is intended to be used with config data in a file. Couldn't we use something like this, which takes case about up to date config etc? We, most likely do not need to dump the whole CS.

I would keep the CEs section of the Synchroniser as well as no CS related stuff (pilot files). Then we just need to dump `Operations` and get the `VO` specific section in `pilotTools.` So the CS related part of the Synchroniser would look like this:

```python
import datetime
from DIRAC import gLogger, gConfig, S_OK
import json

class PilotCStoJSONSynchronizer:
    def __init__(self):
        """c'tor                                                                                                                                                                                                                        
        Just setting defaults                                                                                                                                                                                                           
        """
        self.log = gLogger.getSubLogger(__name__)

    def getCSDict(self, includeMasterCS=True):

        self.log.info("-- Getting the content of the CS --")
        pilotDict = {
            "timestamp": datetime.datetime.utcnow().isoformat(),
            "CEs": {},
            "GenericPilotDNs": [],
        }
        opRes = gConfig.getOptionsDictRecursively("/Operations")
        if not opRes["OK"]:
            self.log.error("Can't get sections from Operations", opRes["Message"])
            return opRes
        return pilotDict | opRes['Value']

if __name__ == "__main__":
   from DIRAC.Core.Base.Script import parseCommandLine
   parseCommandLine()
   client = PilotCStoJSONSynchronizer()
   res = client.getCSDict()

   print(json.dumps(res, indent = 4))
```

Reply:
If that works, fine for me.

Reply:
I've just implemented in the existing synchroniser. It works fine, however we have to be aware that it dumps a whole `Operations` section, not only its `pilot` subsections.

Reply:
One more question on `pilotTools`. When modifying the code to handle a new JSON format I noticed that the current implementation can handle both a comma separated string for a list-like data (i.e. pilot commands) and the proper Python lists. Now, from what I can see above, and my CS to JSON dump code I have already written, the  commands are stored in a string : `"command1, command2, etc"` . Since we are dealing with a new format can we decide on coding only for the string case (this is what `diracfg` emits). This would avoid that terrible _if-then-elsery_.